### PR TITLE
pose bottomup higherhrnet: data

### PR DESCRIPTION
**Pull request by zhiboniu**
pose bottomup higherhrnet: data

### CODE REVIEW

File: ppdet/data/source/keypoint_bottomup_coco.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
+# Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline ppdet/data/source/keypoint_bottomup_coco.py]: 2021

File: ppdet/data/source/keypoint_bottomup_coco.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+import copy
+
+import xtcocotools
+from xtcocotools.coco import COCO
+# from paddle.io import Dataset
+from .dataset import DetDataset
+from ppdet.core.workspace import register, serializable
+
+@serializable
+class Keypoint_BottomUpBaseDataset(DetDataset):
```
Comment (qingqing01) [inline ppdet/data/source/keypoint_bottomup_coco.py]: 命名，没有下划线

线下讨论是讲 bottomup和topdown的dataset合并吧

File: ppdet/data/source/keypoint_bottomup_coco.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+# limitations under the License.
+
+import os
+import cv2
+import numpy as np
+import copy
+
+import xtcocotools
+from xtcocotools.coco import COCO
+# from paddle.io import Dataset
```
Comment (qingqing01) [inline ppdet/data/source/keypoint_bottomup_coco.py]: remove unsed code

File: ppdet/data/source/keypoint_bottomup_coco.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+# See the License for the specific language governing permissions and   
+# limitations under the License.
+
+import os
+import cv2
+import numpy as np
+import copy
+
+import xtcocotools
+from xtcocotools.coco import COCO
```
Comment (qingqing01) [inline ppdet/data/source/keypoint_bottomup_coco.py]: 这里用xtcocotools而不是pycocotools?

File: ppdet/data/transform/keypoint_operators.py [hunk 5]
```diff
@@ -0,0 +1,462 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline ppdet/data/transform/keypoint_operators.py]: 2021

File: ppdet/data/transform/keypoint_operators.py [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+logger = setup_logger(__name__)
+
+registered_ops = []
+
+__all__ = ['RandomAffine', 'KeyPoint_Flip', 'TagGenerate', 'ToHeatmaps', 'Normalize_Permute', 'EvalAffine', 'get_affine_transform', 'transpred']
+
+def register_keypointop(cls):
+    return serializable(cls)
+
+class ImageError(ValueError):
```
Comment (qingqing01) [inline ppdet/data/transform/keypoint_operators.py]: remove unsed code

File: ppdet/data/transform/keypoint_operators.py [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+    pass
+
+@register_keypointop
+class KeyPoint_Flip(object):
```
Comment (qingqing01) [inline ppdet/data/transform/keypoint_operators.py]: 1. KeyPoint_Flip -> KeyPointFlip
2. add comments
3. whether need to inherit BaseOperator in  transform/operators.py

File: ppdet/data/transform/keypoint_operators.py [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+        assert isinstance(flip_permutation, Sequence)
+        self.flip_permutation = flip_permutation
+        self.flip_prob = flip_prob
+        self.hmsize = hmsize
+
+    def __call__(self, records):
+        image = records['image']
+        kpts_lst = records['joints']
+        mask_lst = records['mask']
+        #np.random.seed(0)   #debug
```
Comment (qingqing01) [inline ppdet/data/transform/keypoint_operators.py]: remove debug code

File: ppdet/data/transform/keypoint_operators.py [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+
+    def __call__(self, records):
+        image = records['image']
+        keypoints = records['joints']
+        heatmap_mask = records['mask']
+        #np.random.seed(0)   #debug
+        degree = (np.random.random() * 2 - 1) * self.max_degree
+        shape = np.array(image.shape[:2][::-1])
+        center = center = np.array((np.array(shape) / 2))
+        #np.random.seed(0)   #debug
```
Comment (qingqing01) [inline ppdet/data/transform/keypoint_operators.py]: remove debug code in line 268 and line 272

File: ppdet/data/transform/keypoint_operators.py [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+        if self.is_scale:
+            image /= 255.
+        image = image.transpose((2, 0, 1))
+        mean = np.array(self.mean, dtype=np.float32)
+        std = np.array(self.std, dtype=np.float32)
+        invstd = 1. / std
+        for v, m, s in zip(image, mean, invstd):
+            v.__isub__(m).__imul__(s)
+        records['image'] = image
+        return records
```
Comment (qingqing01) [inline ppdet/data/transform/keypoint_operators.py:311]: 可以复用 ./transform/operators.py 中的 Normalize和Permute吧

**Review by qingqing01 — COMMENTED** [review]
Remove model config in this PR

File: ppdet/data/source/keypoint_bottomup_coco.py (hunk 1 above)
Comment (zhiboniu) [reply ppdet/data/source/keypoint_bottomup_coco.py]: done

File: ppdet/data/source/keypoint_bottomup_coco.py (hunk 2 above)
Comment (zhiboniu) [reply ppdet/data/source/keypoint_bottomup_coco.py]: done

File: ppdet/data/source/keypoint_bottomup_coco.py (hunk 3 above)
Comment (zhiboniu) [reply ppdet/data/source/keypoint_bottomup_coco.py]: done

File: ppdet/data/source/keypoint_bottomup_coco.py (hunk 4 above)
Comment (zhiboniu) [reply ppdet/data/source/keypoint_bottomup_coco.py]: 因为还用到了crowdpose数据集，xtcocotools兼容pycocotools和crowdpose，所以可以通用。后面可以提个pr把pycocotools都换成xtcocotools

File: ppdet/data/transform/keypoint_operators.py (hunk 5 above)
Comment (zhiboniu) [reply ppdet/data/transform/keypoint_operators.py]: done

File: ppdet/data/transform/keypoint_operators.py (hunk 6 above)
Comment (zhiboniu) [reply ppdet/data/transform/keypoint_operators.py]: done

File: ppdet/data/transform/keypoint_operators.py (hunk 7 above)
Comment (zhiboniu) [reply ppdet/data/transform/keypoint_operators.py]: done,
re 3. not too much codes, rewrite is clear to read

File: ppdet/data/transform/keypoint_operators.py (hunk 8 above)
Comment (zhiboniu) [reply ppdet/data/transform/keypoint_operators.py]: done

File: ppdet/data/transform/keypoint_operators.py (hunk 9 above)
Comment (zhiboniu) [reply ppdet/data/transform/keypoint_operators.py]: done

File: ppdet/data/transform/keypoint_operators.py (hunk 10 above)
Comment (zhiboniu) [reply ppdet/data/transform/keypoint_operators.py:311]: 也可以，先保留吧，两种方式都可以用

File: ppdet/data/reader.py [hunk 11]
```diff
@@ -25,7 +25,7 @@
 import numpy as np
 
 from paddle.io import DataLoader, DistributedBatchSampler
-from paddle.fluid.dataloader.collate import default_collate_fn
+from paddle.fluid.reader import default_collate_fn
```
Comment (qingqing01) [inline ppdet/data/reader.py]: doc in offical website is :   paddle.io. default_collate_fn ( ) 

https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/io/default_collate_fn_cn.html#default-collate-fn

@heavengate Please help to check

File: ppdet/data/source/keypoint_coco.py [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  
+# See the License for the specific language governing permissions and   
+# limitations under the License.
+
+import os
+import cv2
+import numpy as np
+import copy
+
+import xtcocotools
```
Comment (qingqing01) [inline ppdet/data/source/keypoint_coco.py:20]: please add todo for  unify xtcocotools and pycocotools . If changed also need to update https://github.com/PaddlePaddle/PaddleDetection/blob/develop/requirements.txt

Comment (zhiboniu) [reply ppdet/data/source/keypoint_coco.py:20]: done

**Review by qingqing01 — DISMISSED** [review]

File: ppdet/data/reader.py (hunk 11 above)
Comment (zhiboniu) [reply ppdet/data/reader.py]: 已确认，paddle最新develop版本无需做此修改。已恢复

**Review by qingqing01 — APPROVED** [review]

**Comment by innerlee** [comment]
Hi dear maintainers,

Please add proper acknowledgement and the original licence and copyright notice, if codes are modified from mmpose. See https://github.com/open-mmlab/mmpose/blob/master/LICENSE

**Comment by zhiboniu** [comment]
> Hi dear maintainers,
> 
> Please add proper acknowledgement and the original licence and copyright notice, if codes are modified from mmpose. See https://github.com/open-mmlab/mmpose/blob/master/LICENSE

Thanks for the reminder, mmpose is an awesome job. I will add the notice in later work soon!

**Pull request merged**